flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -271,12 +271,28 @@ public int getTotalNumberOfMemorySegments() {
 		return totalNumberOfMemorySegments;
 	}
 
+	public long getTotalMemory() {
+		return getTotalNumberOfMemorySegments() * memorySegmentSize;
+	}
+
 	public int getNumberOfAvailableMemorySegments() {
 		synchronized (availableMemorySegments) {
 			return availableMemorySegments.size();
 		}
 	}
 
+	public long getAvailableMemory() {
+		return getNumberOfAvailableMemorySegments() * memorySegmentSize;
+	}
+
+	public int getNumberOfUsedMemorySegments() {
+		return getTotalNumberOfMemorySegments() - getNumberOfAvailableMemorySegments();
+	}
+
+	public long getUsedMemory() {
+		return getNumberOfUsedMemorySegments() * memorySegmentSize;
+	}
```

Comment:
What is the semantic of this method? Shall it return the number of handed out memory segments to a set of `LocalBufferPools` or shall it return the size of the actually used memory? If it is the latter, then I am afraid that memory which has been handed out to the `LocalBufferPools` might also be idling there. Hence this value won't tell us how much memory is currently been used.

Reply:
Good catch. I didn't consider that. I'm gonna look into accessing the `LocalBufferPools`.

Reply:
I discussed the issue with @NicoK. He objects looking for the memory usage in the `LocalBufferPools` and the `BufferManager`, i.e. the exclusive buffers, since these values are configured separately using `taskmanager.network.memory.buffers-per-channel` and `taskmanager.network.memory.floating-buffers-per-gate`. I think, that’s a fair point. Exposing only the amount of memory acquired by the non-available `MemorySegments` in the `NetworkBufferPool` should be good enough. This is what is actually configured through `taskmanager.memory.network.min` and `taskmanager.memory.network.max`. Any thoughts about that?

Reply:
I tend to agree with @NicoK. As I understand, we want to expose the 'used' metric per TM. In that sense, if some buffers are already reserved for a certain task then they cannot be used for others, so they are not available on TM level. The 'used' buffers (containing records to process, I guess) per task/channel should be a metric on task/channel level, otherwise it is hard to judge what 'available' actually means.

Reply:
also +1 on my side. 
I might previously misunderstand the motivation of exposing the TM level buffer usages. 

I guess this metric is for guiding the users how to tune the config of TM network buffers:
- If the config of TM network buffers is not enough, the task will fail when deployment.
- If the config of buffer amount is redundant, we can know this info from this proposed metrics when all the task running.

If so, it is indeed not need to care about the internal buffer usages on task level.

Reply:
I see your point. However, `LocalBufferPools` can contain excess memory which can be claimed by other `LocalBufferPools`. See `LocalBufferPool.returnExcessMemorySegments` for more details. Hence it is not correct that memory which is assigned to a `LocalBufferPool` won't be available for other buffer pools.

Ideally the used network memory metric would allow us to see if we have overprovisioned a TM with network memory or whether it it running at its capacity. However, this won't be visible if we report how many memory segments have been handed out by the `NetworkBufferPool` to the `LocalBufferPools` where they might lie around idling.

If there is no easy way to achieve this goal, then I would be ok with keeping as it is which gives a more coarse grained overview over the resource utilization which is already a plus.

Reply:
Ok, I did not release that it is more sophisticated internally.
One would need to understand then what the exact metric is in this case to calculate it.

After offline discussion with @tillrohrmann, we decided to merge this PR to deliver the coarse grained metric.
Later we can revisit this topic if this inaccuracy becomes a bigger problem/need for users.